plugins/ftdidmx/FtdiDmxThread.cpp

```diff
@@ -73,15 +97,133 @@ bool FtdiDmxThread::WriteDMX(const DmxBuffer &buffer) {
+                    ola::rdm::RDMCallback *callback) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if(m_pending_request == nullptr) {
+      m_pending_request = request;
+      m_rdm_callback = callback;
+  } else {
+    OLA_WARN << "Unable to queue RDM request, RDM operation already pending";
+  }
+  OLA_WARN << "Function not properly implemented yet, callback may not get called.";
+}
+
+void FtdiDmxThread::RunFullDiscovery(ola::rdm::RDMDiscoveryCallback *callback) {
+  m_discovery_agent.StartFullDiscovery(ola::NewSingleCallback(this, &FtdiDmxThread::DiscoveryComplete, callback));
+}
+
+void FtdiDmxThread::RunIncrementalDiscovery(ola::rdm::RDMDiscoveryCallback *callback) {
+  m_discovery_agent.StartIncrementalDiscovery(ola::NewSingleCallback(this, &FtdiDmxThread::DiscoveryComplete, callback));
+}
+
+/**
+ * Called when the discovery process finally completes
+ * @param callback the callback passed to StartFullDiscovery or
+ * StartIncrementalDiscovery that we should execute.
+ * @param status true if discovery worked, false otherwise
+ * @param uids the UIDSet of UIDs that were found.
+ */
+void FtdiDmxThread::DiscoveryComplete(ola::rdm::RDMDiscoveryCallback *callback,
+                                       bool,
+                                       const ola::rdm::UIDSet &uids) {
+  OLA_DEBUG << "FTDI discovery complete: " << uids;
+  if (callback) {
+    callback->Run(uids);
+  }
+}
+
+/**
+ * @brief Method called to cleanup any outstanding callbacks
+ * @param state
+ *
+ * All callbacks except the RDMCallback lack a way of reporting an error state to the caller.
+ */
+void FtdiDmxThread::destroyPendindingCallback(ola::rdm::RDMStatusCode state) {
+  MuteDeviceCallback *thread_mute_callback = nullptr;
+  UnMuteDeviceCallback *thread_unmute_callback = nullptr;
+  BranchCallback *thread_branch_callback = nullptr;
+  ola::rdm::RDMCallback *thread_rdm_callback = nullptr;
+
+  if(m_mute_complete != nullptr) {
+    thread_mute_callback = m_mute_complete;
+    m_mute_complete = nullptr;
+    thread_mute_callback->Run(false);
+  } else if(m_unmute_complete != nullptr) {
+    thread_unmute_callback = m_unmute_complete;
+    m_unmute_complete = nullptr;
+    thread_unmute_callback->Run();
+  } else if(m_branch_callback != nullptr) {
+    thread_branch_callback = m_branch_callback;
+    m_branch_callback = nullptr;
+    thread_branch_callback->Run(nullptr, 0);
+  } else if(m_rdm_callback != nullptr) {
+    thread_rdm_callback = m_rdm_callback;
+    m_rdm_callback = nullptr;
+    ola::rdm::RunRDMCallback(thread_rdm_callback, state);
+  }
+}
+
+void FtdiDmxThread::MuteDevice(const ola::rdm::UID &target,
+                               MuteDeviceCallback *mute_complete) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if(m_pending_request == nullptr) {
+    OLA_INFO << "Muting device";
+    m_mute_complete = mute_complete;
+    m_pending_request = ola::rdm::NewMuteRequest(m_uid, target, m_transaction_number += 1);
```

Comment:
Will m_transaction_number++ not work here?

Reply:
Absolutely.
I'm from the school of thought that says that using += is better because it prevents situations where you have 2 semi successive lines doing ++ (for bad reasons) which then go missed in debugging.

Reply:
Unless you are opposed to this style I'm leaving it.

Reply:
I'd prefer to stick with ++ given it's what we've got across the whole of the rest of the code-base. I'm not really clear how += protects from having multiple lines doing the same thing (I assume you were referring to a copy/paste?).

Reply:
I don't remember where I heard it but I'll try to fin a reference, I was told in the past that secure and clear coding should prefer += over ++

Reply:
Okay, I'd still suggest that being consistent within the code-base is probably more valuable, but perhaps updating everything to += should be saved for another PR. I know someone else prefers ++foo as it's more efficient apparently (assuming the compiler doesn't optimise it away).

Reply:
I switched to ++ but have them outside the new() operation as a distinct step so that mistakes don't happen.